hibernate-core/src/test/java/org/hibernate/orm/test/annotations/generics/EmbeddedIdGenericsSuperclassTest.java

```diff
@@ -0,0 +1,279 @@
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.orm.test.annotations.generics;
+
+import java.io.Serializable;
+import java.util.Random;
+
+import org.hibernate.testing.orm.junit.DomainModel;
+import org.hibernate.testing.orm.junit.Jira;
+import org.hibernate.testing.orm.junit.SessionFactory;
+import org.hibernate.testing.orm.junit.SessionFactoryScope;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import jakarta.persistence.Embeddable;
+import jakarta.persistence.EmbeddedId;
+import jakarta.persistence.Entity;
+import jakarta.persistence.MappedSuperclass;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Path;
+import jakarta.persistence.criteria.Root;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Marco Belladelli
+ */
+@DomainModel(annotatedClasses = {
+		EmbeddedIdGenericsSuperclassTest.Customer.class,
+		EmbeddedIdGenericsSuperclassTest.Invoice.class,
+		EmbeddedIdGenericsSuperclassTest.NestedCustomer.class
+})
+@SessionFactory
+@Jira("https://hibernate.atlassian.net/browse/HHH-16188")
+public class EmbeddedIdGenericsSuperclassTest {
+	@BeforeAll
+	public void setUp(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final Customer customer = new Customer( 1, "Francisco" );
+			session.persist( customer );
+			final Invoice invoice = new Invoice( 2 );
+			session.persist( invoice );
+			final NestedCustomer nestedCustomer = new NestedCustomer( "Marco" );
+			session.persist( nestedCustomer );
+		} );
+	}
+
+	@AfterAll
+	public void tearDown(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			session.createMutationQuery( "delete from Customer" ).executeUpdate();
+			session.createMutationQuery( "delete from Invoice" ).executeUpdate();
+			session.createMutationQuery( "delete from NestedCustomer" ).executeUpdate();
+		} );
+	}
+
+	@Test
+	public void testCustomer(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final Customer customer = session.createQuery(
+					"from Customer c where c.id.someDomainField = 1",
+					Customer.class
+			).getSingleResult();
+			assertThat( customer ).isNotNull();
+			assertThat( customer.getCode() ).isEqualTo( 1 );
+			assertThat( customer.getName() ).isEqualTo( "Francisco" );
+		} );
+	}
+
+	@Test
+	public void testCustomerCriteria(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final CriteriaBuilder cb = session.getCriteriaBuilder();
+			final CriteriaQuery<Customer> query = cb.createQuery( Customer.class );
+			final Root<Customer> root = query.from( Customer.class );
+			final Path<CustomerId> id = root.get( "id" );
+			assertThat( id ).isNotNull();
+			assertThat( id.getJavaType() ).isEqualTo( CustomerId.class );
```

Comment:
Not sure if this is "correct". IMO the exposed type here should rather still be `DomainEntityId`, hence my comment about doing the generic lookup in `SqmEmbeddedValuedSimplePath#get` instead to avoid this exposure.

Reply:
Can you also please add an assertion for `id.getModel()`?

Reply:
Oh sorry, I interpreted *don't expose any generic info* as we should return the actual concrete type instead of the generic one, but I see now what you meant.

Then this is indeed not correct, I'm gonna move back the logic to `SqmEmbeddedValuedSimplePath` and fix the assertions.

Reply:
Wait a second. I'm checking the spec now to be sure.

Reply:
Updated with:
* handling of generic embedded non-identifier properties
* review of `getModel()` logic for `SingularAttributeImpl`

The new logic required some impl for `getNodeType()` and `getReferencedPathSource()` (see `AbstractSqmSimplePath`). I also added the new `JpaPath#getResolvedModel` as suggested.